pkg/controller/certificates/sync.go

```diff
@@ -114,20 +125,48 @@ func (c *Controller) Sync(ctx context.Context, crt *v1alpha1.Certificate) (err e
 		return err
 	}
 
+	// update certificate expiry metric
+	defer c.updateCertificateExpiryMetric(crt)
```

Comment:
By doing this here, we perform two lookups against the lister and also decode the pem/certificate twice.

I'd rather use the `cert.NotAfter` further down (https://github.com/dippynark/cert-manager/blob/0098c6976354d6993e1e48307ab0d249cc7f689d/pkg/controller/certificates/sync.go#L178) for this, to save extra calls.

If the reasoning for this is so that we update with the correct expiry time even after a renewal, +1 however I don't think this actually *will* behave as you wish, as the lister is a *cache* that is populated by the Informer, and it's unlikely the informer will have seen the change yet here.

Reply:
It's there just so that it's called before any certificate renewal, otherwise sync might return without ever updating the certificate expiry date

Reply:
Let's replace this with `defer c.metrics.UpdateCertificateExpiry(crt)`